### PR TITLE
Allow nested routing for objects related to repos

### PR DESCRIPTION
**Pull request by asmacdo**
This work allows objects that are related to repositories to have urls
that are nested within their related repository. For example, an
`rpm` importer named `zoo-importer` related to the `zoo` repository
will live in `repositories/zoo/importers/rpm/zoo-importer/`

### CODE REVIEW

**Comment by seandst** [comment]
I think we'll never merge anything that's got those "Example*" models in there. You might be able to move some things around and incidentally get https://pulp.plan.io/issues/2369 completed.

File: app/pulp/app/models/repository.py [hunk 1]
```diff
@@ -232,6 +232,18 @@ class Meta(ContentAdaptor.Meta):
         default_related_name = 'importers'
 
 
+class ExampleDetailImporter(Importer):
+    # I didn't see this variable being used?
```
Comment (seandst) [inline app/pulp/app/models/repository.py]: Which variable?

You should be setting `TYPE`, not `type`.

File: app/pulp/app/serializers/repository.py [hunk 2]
```diff
@@ -37,3 +38,34 @@ class Meta:
...
+        abstract = True
+        fields = MasterModelSerializer.Meta.fields + ('name',)
+
+
+class ExampleDetailImporterSerializer(ImporterSerializer):
+    """
+    Example of a Detail serializer that also utilizes nested routes.
+    """
+    # Content adapters need to use two kwargs because the name is only unique within a
+    # repository and alone, `Importer.name` is not enough to uniquely identify it.
```
Comment (seandst) [inline app/pulp/app/serializers/repository.py]: This comment doesn't make sense to me in this context. It talks about needing to use two kwargs, but only one kwarg is used on the next line. Probably reword this to be more like "Serializers for ContentAdapter models have a special identity that is only valid in the context of the repositories that they're related to, which RepositoryNestedIdentityField knows how to handle", or something.

Comment (seandst) [inline app/pulp/app/serializers/repository.py]: Also, and again, there's no way any of these "Example*" classes get merge into the platform app code.

File: app/pulp/app/viewsets/base.py [hunk 3]
```diff
@@ -87,11 +106,9 @@ def register_with(cls, router):
                        'correctly subclass the Master ViewSet, and do both have endpoint_name '
                        'set to different values?').format(cls.__name__)
                 warnings.warn(msg, RuntimeWarning)
```
Comment (seandst) [inline app/pulp/app/viewsets/base.py]: You actually got bit by this warning while working on this. Should it be a RuntimeError instead? I'm thinking probably it should.

File: app/pulp/app/viewsets/repository.py [hunk 4]
```diff
@@ -41,3 +44,34 @@ def content(self, request, name):
...
+    @decorators.detail_route()
+    def importers(self, request, name):
+        # Creates a repositories/<repo>/importers/ endpoint.
+        # This will link to all importers that are associated within the repository.
+        repo = self.get_object()
+        importers = Importer.objects.filter(repository__name=repo.name)
+        paginator = UUIDPagination()
+        page = paginator.paginate_queryset(importers, request)
+        serializer = ImporterSerializer(page, many=True, context={'request': request})
+        return paginator.get_paginated_response(serializer.data)
```
Comment (seandst) [inline app/pulp/app/viewsets/repository.py]: This is so similar to the content detail route that it would be useful to have a sort of factory function to generalize the combination of a queryset, paginator class, and serializer class into a paginated response.

File: app/pulp/app/viewsets/repository.py [hunk 5]
```diff
@@ -41,3 +44,34 @@ def content(self, request, name):
...
+    # Indicates that importer urls should be routed through their associated repository.
+    nested_parent = RepositoryViewSet
+    # Name the parameter to generate the link. This is important because `name` in this url refers
+    # to `Importer.name`, not `Repository.name`
+    nested_parent_lookup_name = 'repo_name'
+    endpoint_name = 'importers'
+    serializer_class = ImporterSerializer
+
+
+class ExampleDetailImporterViewSet(ImporterViewSet):
```
Comment (seandst) [inline app/pulp/app/viewsets/repository.py]: 'nother one that'll never get merged to the platform app.

**Comment by asmacdo** [comment]
Yeah, the example model/serializer/viewset aren't intended to be merged either, but I felt that they were necessary to show how the other changes will be used.

Good idea about 2369, though I assumed that you probably had a more complete working example than what I have here. If not, lets chat about how that might work.

**Comment by seandst** [comment]
I mainly focused on the loading code, which proved to be a minefield frought with peril. The branch I linked should hopefully be pretty easy to merge what what you've created here by moving the example stuff over into that framework.

I didn't comment my branch much at all, just committed it as soon as it didn't explode and moved on.

File: app/pulp/app/viewsets/base.py (hunk 3 above)
Comment (asmacdo) [inline app/pulp/app/viewsets/base.py]: Yeah that makes sense. I don't see a benefit to letting Pulp be successful with missing endpoints. We want to make sure a problem like that fails obviously.

**Comment by asmacdo** [comment]
@seandst, the non-production code has been removed, ready for review.

**Comment by seandst** [comment]
LGTM, but travis is indicating a problem with the viewset registration mechanism.

**Comment by asmacdo** [comment]
@seandst, travis fix was a 1 line change, I had to add a queryset to the Importer ViewSet.
